packages/tpetra/core/src/kokkos_refactor/Tpetra_KokkosRefactor_Details_MultiVectorDistObjectKernels.hpp

```diff
@@ -148,6 +148,17 @@ outOfBounds (const IntegerType x, const IntegerType exclusiveUpperBound)
         ("Tpetra::MultiVector pack one col",
          range_type (0, idx.size ()),
          PackArraySingleColumn (dst, src, idx, col));
+
+      // FENCE REVIEW - CONFIRMED FAILURE
+      //   Testing: This code is exercised by unit tests.
+      //   GTX960:  Passed (confirmed) with CUDA_LAUNCH_BLOCKING=0 and fence removed.
+      //   White:   Failed with CUDA_LAUNCH_BLOCKING=0 and fence removed.
+          // TpetraCore_CrsMatrix_ReplaceDomainMapAndImporter_MPI_4
+          // TpetraCore_ImportExport2_UnitTests_MPI_4
+          // TpetraCore_MatrixMatrix_UnitTests_MPI_4
+      //   Plan:    Need to determine downstream failure point. Fix and remove this fence.
+      //   Notes:
+
```

Comment:
On `ride` with `TPETRA_DEBUG=ON` and `TPETRA_DEBUG=OFF` I did not see failures in these three tests with the fence is removed.

Reply:
Oh, wait, didn't set `CUDA_LAUNCH_BLOCKING=0`, I'll rerun.

Reply:
Even with `CUDA_LAUNCH_BLOCKING=0`, these still pass.  Others fail, though.

Reply:
I confirmed the failures on white for removing this fence. However note the 2nd fence you asked about below I can not replicate the failures. So that may have been an error in my recording.

Logged into nodes like this: `bsub -Is -n 16 -q rhel7G bash`

Configure:
```
cmake \
  -DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/atdm/ATDMDevEnv.cmake \
  -DTpetra_ENABLE_TESTS=ON -DTrilinos_ENABLE_Tpetra=ON \
  ../trilinos
```

For this current branch state.
Running with CUDA_LAUNCH_BLOCKING=1

	 52 - TpetraCore_Core_initialize_where_tpetra_initializes_mpi_and_user_initializes_kokkos_MPI_2 (Failed)

Running with CUDA_LAUNCH_BLOCKING=0

	 44 - TpetraCore_Core_initialize_where_user_initializes_mpi_MPI_4 (Failed)
	 45 - TpetraCore_Core_ScopeGuard_where_user_initializes_mpi_MPI_4 (Failed)
	 51 - TpetraCore_Core_ScopeGuard_where_user_initializes_kokkos_MPI_1 (Failed)
	 53 - TpetraCore_Core_ScopeGuard_where_tpetra_initializes_mpi_and_user_initializes_kokkos_MPI_2 (Failed)

Remove this fence with CUDA_LAUNCH_BLOCKING=0:

	 44 - TpetraCore_Core_initialize_where_user_initializes_mpi_MPI_4 (Failed)
	 45 - TpetraCore_Core_ScopeGuard_where_user_initializes_mpi_MPI_4 (Failed)
	 51 - TpetraCore_Core_ScopeGuard_where_user_initializes_kokkos_MPI_1 (Failed)
	 53 - TpetraCore_Core_ScopeGuard_where_tpetra_initializes_mpi_and_user_initializes_kokkos_MPI_2 (Failed)
	 81 - TpetraCore_CrsMatrix_ReplaceDomainMapAndImporter_MPI_4 (Failed)
	130 - TpetraCore_ImportExport2_UnitTests_MPI_4 (Failed)
	165 - TpetraCore_MatrixMatrix_UnitTests_MPI_4 (Failed)

So that confirmed the listed failures and also that shows the extra fails we discussed in meeting which I have been ignoring for the moment. Soon I'll be rebasing this branch to the latest develop to incorporate a bunch of relevant PRs and I'll update the comments.

Reply:
This fence is in the pack function. Examining other pack functions, no fence is applied at the end. Following the design rule (apply fence as late as possible and tpetra api leave it open-end), the fence should be called from the caller of pack function in my opinion.